### PR TITLE
PYTHON-4005 Replace flake8 and isort with ruff

### PR DESCRIPTION
**Pull request by blink1073**
I disabled many of the rules that would have caused extra work, especially in `tests`.  We can consider re-enabling them over time.  I used https://learn.scientific-python.org/development/guides/style/#ruff as the basis for the configuration.

It automatically added `from __future__ import annotations`, which accounts for the high file changed count.

### CODE REVIEW

File: pymongo/operations.py [hunk 1]
```diff
@@ -547,7 +547,7 @@ def document(self) -> dict[str, Any]:
 class SearchIndexModel:
     """Represents a search index to create."""
 
-    __slots__ = "__document"
+    __slots__ = ("__document",)
```
Comment (blink1073) [inline pymongo/operations.py:550]: Found a bug!

File: pymongo/monitoring.py [hunk 2]
```diff
@@ -1216,7 +1212,7 @@ class ServerClosedEvent(_ServerEvent):
 class TopologyEvent:
     """Base class for topology description events."""
 
-    __slots__ = "__topology_id"
+    __slots__ = ("__topology_id",)
```
Comment (blink1073) [inline pymongo/monitoring.py:1215]: Found a bug!

File: test/__init__.py [hunk 3]
```diff
@@ -795,11 +796,12 @@ def is_topology_type(self, topologies):
             return True
         return False
 
-    def require_cluster_type(self, topologies=[]):  # noqa
+    def require_cluster_type(self, topologies=None):
```
Comment (blink1073) [inline test/__init__.py:799]: Found a bug!

File: doc/docs-requirements.txt [hunk 4]
```diff
@@ -1,5 +1,5 @@
-sphinx>=5.3,<7
-sphinx_rtd_theme~=0.5
-readthedocs-sphinx-search~=0.1
+sphinx>=7,<8
```
Comment (ShaneHarvey) [inline doc/docs-requirements.txt:1]: Is this related to ruff?

File: gridfs/grid_file.py [hunk 5]
```diff
@@ -357,12 +357,14 @@ def write(self, data: Any) -> None:
         except AttributeError:
             # string
             if not isinstance(data, (str, bytes)):
-                raise TypeError("can only write strings or file-like objects")
+                raise TypeError("can only write strings or file-like objects") from None
```
Comment (ShaneHarvey) [inline gridfs/grid_file.py:360]: What are the pros of adding `from None`? Are there any cons?

File: test/__init__.py (hunk 3 above)
Comment (ShaneHarvey) [reply test/__init__.py:799]: Was there a bug here? I don't see it. `topologies` is read only IIUC so the mutable default arg doesn't have and effect.

File: pymongo/monitoring.py (hunk 2 above)
Comment (ShaneHarvey) [reply pymongo/monitoring.py:1215]: Also not a bug.

File: pymongo/operations.py (hunk 1 above)
Comment (ShaneHarvey) [reply pymongo/operations.py:550]: Not a bug :). `__slots__` can be assigned a single attr name or a tuple of names: https://docs.python.org/3/reference/datamodel.html#object.__slots__

File: bson/__init__.py [hunk 6]
```diff
@@ -244,7 +244,7 @@ def _raise_unknown_type(element_type: int, element_name: str) -> NoReturn:
 
 
 def _get_int(
-    data: Any, view: Any, position: int, dummy0: Any, dummy1: Any, dummy2: Any
+    data: Any, _view: Any, position: int, dummy0: Any, dummy1: Any, dummy2: Any
```
Comment (ShaneHarvey) [inline bson/__init__.py:247]: Why did ruff rename "view" -> "_view"? It seems out of place.

Comment (blink1073) [reply bson/__init__.py:247]: I did that manually, `dummy` is explicitly part of `dummy-variable-rgx`.

File: pymongo/uri_parser.py [hunk 7]
```diff
@@ -598,14 +598,14 @@ def _parse_kms_tls_options(kms_tls_options: Optional[Mapping[str, Any]]) -> dict
     if not isinstance(kms_tls_options, dict):
         raise TypeError("kms_tls_options must be a dict")
     contexts = {}
-    for provider, opts in kms_tls_options.items():
-        if not isinstance(opts, dict):
+    for provider, options in kms_tls_options.items():
```
[the remark this answers is not in the document]
Comment (blink1073) [reply pymongo/uri_parser.py:601]: It was shadowing the function parameter

File: gridfs/grid_file.py (hunk 5 above)
Comment (blink1073) [reply gridfs/grid_file.py:360]: It affects whether the previous error shows up in the traceback or not.

File: pymongo/operations.py (hunk 1 above)
Comment (blink1073) [reply pymongo/operations.py:550]: Ah, shucks.

File: test/__init__.py (hunk 3 above)
Comment (blink1073) [reply test/__init__.py:799]: Yeah fair, but nice to see that it catches this case.

**Review by ShaneHarvey — APPROVED** [review]
LGTM!

**Comment by blink1073** [comment]
cc @charliermarsh 🚀

**Comment by charliermarsh** [comment]
Awesome to see this, thanks @blink1073!
